File: src/analyzer/preprocessor.js
export class SentenceBlock {
  constructor(text, startChar, endChar) {
    this.text = text;
    this.start_char = startChar;
    this.end_char = endChar;
  }
}

export class ParagraphBlock {
  constructor(text, startChar, endChar, isHeading, isListItem, headingLevel, sentences) {
    this.text = text;
    this.start_char = startChar;
    this.end_char = endChar;
    this.is_heading = isHeading;
    this.is_list_item = isListItem;
    this.heading_level = headingLevel;
    this.sentences = sentences;
  }
}

export class DocumentBlock {
  constructor(rawText, normalizedText, paragraphs) {
    this.raw_text = rawText;
    this.normalized_text = normalizedText;
    this.paragraphs = paragraphs;
  }
}

export class Preprocessor {
  constructor(language = 'auto') {
    this.language = language;
  }

  normalizeText(text) {
    let normalized = text.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
    if (typeof normalized.normalize === 'function') {
      normalized = normalized.normalize('NFKC');
    }
    return normalized;
  }

  process(text) {
    const normalizedText = this.normalizeText(text);
    const paragraphs = this._splitParagraphs(normalizedText);
    return new DocumentBlock(text, normalizedText, paragraphs);
  }

  _splitParagraphs(text) {
    const paragraphs = [];
    const lines = text.split('\n');

    let currentPLines = [];
    let charOffset = 0;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const lineLen = line.length;
      const lineStart = charOffset;
      const lineEnd = charOffset + lineLen;

      // Advance charOffset for next line (+1 for '\n')
      charOffset = lineEnd + 1;

      const stripped = line.trim();
      if (!stripped) {
        if (currentPLines.length > 0) {
          const pBlock = this._createParagraphBlock(currentPLines, text);
          if (pBlock) paragraphs.push(pBlock);
          currentPLines = [];
        }
        continue;
      }

      const headingMatch = stripped.match(/^(#{1,6})\s+(.*)$/);
      const listMatch = stripped.match(/^(?:[-*+・]|\d+\.)\s+(.*)$/);

      if (headingMatch || listMatch) {
        if (currentPLines.length > 0) {
          const pBlock = this._createParagraphBlock(currentPLines, text);
          if (pBlock) paragraphs.push(pBlock);
          currentPLines = [];
        }

        const headingLevel = headingMatch ? headingMatch[1].length : 0;
        const isHeading = !!headingMatch;
        const isListItem = !!listMatch;

        const sentences = this._splitSentences(stripped, lineStart);
        paragraphs.push(
          new ParagraphBlock(
            stripped,
            lineStart,
            lineEnd,
            isHeading,
            isListItem,
            headingLevel,
            sentences
          )
        );
      } else {
        currentPLines.push([line, lineStart, lineEnd]);
      }
    }

    if (currentPLines.length > 0) {
      const pBlock = this._createParagraphBlock(currentPLines, text);
      if (pBlock) paragraphs.push(pBlock);
    }

    return paragraphs;
  }

  _createParagraphBlock(linesInfo, fullText) {
    const pText = linesInfo.map(info => info[0]).join('\n');
    const startChar = linesInfo[0][1];
    const endChar = linesInfo[linesInfo.length - 1][2];
    const sentences = this._splitSentences(pText, startChar);
    return new ParagraphBlock(pText, startChar, endChar, false, false, 0, sentences);
  }

  _splitSentences(text, baseOffset) {
    const sentences = [];
    if (!text) return sentences;

    const pattern = /([^。！？.!?]+[。！？.!?]+['”」』\)]*|[^。！？.!?]+$)/g;
    let match;
    let lastIndex = 0;

    while ((match = pattern.exec(text)) !== null) {
      const sText = match[0].trim();
      if (sText) {
        const sStart = baseOffset + match.index;
        const sEnd = baseOffset + match.index + match[0].length;
        sentences.push(new SentenceBlock(sText, sStart, sEnd));
      }
      if (pattern.lastIndex === lastIndex) break;
      lastIndex = pattern.lastIndex;
    }

    if (sentences.length === 0) {
      sentences.push(new SentenceBlock(text, baseOffset, baseOffset + text.length));
    }

    return sentences;
  }
}

File: tanuki_checker/preprocessor.py
import re
import unicodedata
from typing import List, Tuple
from dataclasses import dataclass


@dataclass
class SentenceBlock:
    text: str
    start_char: int
    end_char: int


@dataclass
class ParagraphBlock:
    text: str
    start_char: int
    end_char: int
    is_heading: bool
    is_list_item: bool
    heading_level: int
    sentences: List[SentenceBlock]


@dataclass
class DocumentBlock:
    raw_text: str
    normalized_text: str
    paragraphs: List[ParagraphBlock]


class Preprocessor:
    def __init__(self, language: str = "auto"):
        self.language = language

    def normalize_text(self, text: str) -> str:
        """Apply NFKC normalization and standardize newline characters."""
        text = text.replace("\r\n", "\n").replace("\r", "\n")
        normalized = unicodedata.normalize("NFKC", text)
        return normalized

    def process(self, text: str) -> DocumentBlock:
        normalized_text = self.normalize_text(text)
        paragraphs = self._split_paragraphs(normalized_text)
        return DocumentBlock(
            raw_text=text,
            normalized_text=normalized_text,
            paragraphs=paragraphs,
        )

    def _split_paragraphs(self, text: str) -> List[ParagraphBlock]:
        paragraphs: List[ParagraphBlock] = []
        lines = text.split("\n")
        
        current_p_lines: List[Tuple[str, int, int]] = []
        current_start = 0

        char_offset = 0

        for line in lines:
            line_len = len(line)
            line_start = char_offset
            line_end = char_offset + line_len

            # Advance character offset for next line (including '\n')
            char_offset = line_end + 1

            stripped = line.strip()
            if not stripped:
                if current_p_lines:
                    p_block = self._create_paragraph_block(current_p_lines, text)
                    if p_block:
                        paragraphs.append(p_block)
                    current_p_lines = []
                continue

            # Heading check (# ..., ## ...)
            heading_match = re.match(r"^(#{1,6})\s+(.*)$", stripped)
            # List item check (- ..., * ..., 1. ..., ・ ...)
            list_match = re.match(r"^(?:[-*+・]|\d+\.)\s+(.*)$", stripped)

            if heading_match or list_match:
                # Close any preceding paragraph block
                if current_p_lines:
                    p_block = self._create_paragraph_block(current_p_lines, text)
                    if p_block:
                        paragraphs.append(p_block)
                    current_p_lines = []

                # Headings or List items form standalone paragraph blocks
                heading_level = len(heading_match.group(1)) if heading_match else 0
                is_heading = heading_match is not None
                is_list_item = list_match is not None

                sentences = self._split_sentences(stripped, line_start)
                paragraphs.append(
                    ParagraphBlock(
                        text=stripped,
                        start_char=line_start,
                        end_char=line_end,
                        is_heading=is_heading,
                        is_list_item=is_list_item,
                        heading_level=heading_level,
                        sentences=sentences,
                    )
                )
            else:
                current_p_lines.append((line, line_start, line_end))

        if current_p_lines:
            p_block = self._create_paragraph_block(current_p_lines, text)
            if p_block:
                paragraphs.append(p_block)

        return paragraphs

    def _create_paragraph_block(
        self, lines_info: List[Tuple[str, int, int]], full_text: str
    ) -> ParagraphBlock:
        p_text = "\n".join(info[0] for info in lines_info)
        start_char = lines_info[0][1]
        end_char = lines_info[-1][2]
        sentences = self._split_sentences(p_text, start_char)
        return ParagraphBlock(
            text=p_text,
            start_char=start_char,
            end_char=end_char,
            is_heading=False,
            is_list_item=False,
            heading_level=0,
            sentences=sentences,
        )

    def _split_sentences(self, text: str, base_offset: int) -> List[SentenceBlock]:
        """Split text into sentences for Japanese and English, preserving offsets."""
        sentences: List[SentenceBlock] = []
        if not text:
            return sentences

        # Japanese pattern: terminates at 。！？
        # English pattern: terminates at . ! ? followed by space/end of string
        pattern = r"([^。！？.!?]+[。！？.!?]+['”」』\)]*|[^。！？.!?]+$)"
        
        matches = list(re.finditer(pattern, text, re.MULTILINE))
        if not matches:
            sentences.append(
                SentenceBlock(
                    text=text,
                    start_char=base_offset,
                    end_char=base_offset + len(text),
                )
            )
            return sentences

        for m in matches:
            s_text = m.group(0).strip()
            if s_text:
                s_start = base_offset + m.start()
                s_end = base_offset + m.end()
                sentences.append(
                    SentenceBlock(
                        text=s_text,
                        start_char=s_start,
                        end_char=s_end,
                    )
                )

        return sentences
